lib/Sema/ConstraintGraph.cpp

```diff
@@ -467,98 +373,151 @@ void ConstraintGraph::unbindTypeVariable(TypeVariableType *typeVar, Type fixed){
+    TypeVariableType *typeVar, GatheringKind kind,
     llvm::function_ref<bool(Constraint *)> acceptConstraint) {
-  auto &reprNode = (*this)[CS.getRepresentative(typeVar)];
-  auto equivClass = reprNode.getEquivalenceClass();
-  llvm::SmallPtrSet<TypeVariableType *, 4> typeVars;
-  for (auto typeVar : equivClass) {
-    if (!typeVars.insert(typeVar).second)
-      continue;
-
-    for (auto constraint : (*this)[typeVar].getConstraints()) {
-      if (acceptConstraint(constraint))
-        constraints.insert(constraint);
-    }
+  llvm::TinyPtrVector<Constraint *> constraints;
 
-    auto &node = (*this)[typeVar];
+  /// Add constraints for the given adjacent type variable.
+  llvm::SmallPtrSet<TypeVariableType *, 4> typeVars;
+  llvm::SmallPtrSet<Constraint *, 4> visitedConstraints;
+  auto addAdjacentConstraints = [&](TypeVariableType *adjTypeVar) {
+    auto adjTypeVarsToVisit =
+        (*this)[CS.getRepresentative(adjTypeVar)].getEquivalenceClass();
+    for (auto adjTypeVarEquiv : adjTypeVarsToVisit) {
+      if (!typeVars.insert(adjTypeVarEquiv).second)
+        continue;
 
-    // Retrieve the constraints from adjacent bindings.
-    for (auto adjTypeVar : node.getAdjacencies()) {
-      switch (kind) {
-      case GatheringKind::EquivalenceClass:
-        if (!node.getAdjacency(adjTypeVar).FixedBinding)
+      for (auto constraint : (*this)[adjTypeVarEquiv].getConstraints()) {
+        if (!visitedConstraints.insert(constraint).second)
           continue;
-        break;
 
-      case GatheringKind::AllMentions:
-        break;
+        if (acceptConstraint(constraint))
+          constraints.push_back(constraint);
       }
+    }
+  };
 
-      ArrayRef<TypeVariableType *> adjTypeVarsToVisit;
-      switch (kind) {
-      case GatheringKind::EquivalenceClass:
-        adjTypeVarsToVisit = adjTypeVar;
-        break;
-
-      case GatheringKind::AllMentions:
-        adjTypeVarsToVisit
-          = (*this)[CS.getRepresentative(adjTypeVar)].getEquivalenceClass();
-        break;
-      }
-
-      for (auto adjTypeVarEquiv : adjTypeVarsToVisit) {
-        if (!typeVars.insert(adjTypeVarEquiv).second)
-          continue;
-
-        for (auto constraint : (*this)[adjTypeVarEquiv].getConstraints()) {
-          if (acceptConstraint(constraint))
-            constraints.insert(constraint);
+  auto &reprNode = (*this)[CS.getRepresentative(typeVar)];
+  auto equivClass = reprNode.getEquivalenceClass();
+  for (auto typeVar : equivClass) {
+    auto &node = (*this)[typeVar];
+    for (auto constraint : node.getConstraints()) {
+      if (visitedConstraints.insert(constraint).second &&
+          acceptConstraint(constraint))
+        constraints.push_back(constraint);
+
+      // If we want all mentions, visit type variables within each of our
+      // constraints.
+      if (kind == GatheringKind::AllMentions) {
+        for (auto adjTypeVar : constraint->getTypeVariables()) {
```

Comment:
I think the idea behind caching adjacency is based on the observation that we add constraints a lot less often than we gather them (e.g. to determine bindings), this algorithm is going to be less efficient (maybe not really reflecting in the wall time) because it would iterate over the same equivalence class N times based on the number of unique constraints in each node because equivalence class always includes "self".

Reply:
On the other hand, DenseMap in particular is quite heavyweight in terms of memory usage, even when empty, IIRC

Reply:
I wonder if that size is justified vs. having to allocate/grow and destruct sets of constraints every time we gather which is very often... I'm not saying that I'm against this, just thought that this is worse mentioning.

Reply:
I was looking at `gatherConstraints`, and it was a pretty expensive operation both before and after this change, because it is grabbing constraints from a number of different adjacent type variables and uniquing them along the way. This new implementation is doing basically the same work, but instead of getting the type variables from the stored "adjacencies", we get them from the constraints we are already walking. We do the same uniquing of constraints and of type variables.

This simplification has another purpose, which is to make it easier to ignore specific constraints when performing connected components, so that we can handle one-way constraints specially (https://github.com/apple/swift/pull/25983).